### PR TITLE
[infra] - Cover utils/repo_paths and drop the deleted harness/ from stale rationale comments

### PR DESCRIPTION
**Pull request by cgfixit**
## Proposed changes

Two related pieces of `#1367` residue, both about a reader being unable to audit a decision.

**1. `utils/repo_paths.py` had zero test coverage while staying live.**

Its docstring named a drift test that no longer exists:

> A drift test in `tests/test_harness_agent_routes.py` asserts the two functions agree on a matrix of safe and hostile inputs…

That file was removed with the harness console. `repo_paths` was then the **only first-party module with zero mentions anywhere under `tests/`** — while remaining on a live path: `utils/ops_runner.py:43,458` gates `real-repo-run --read-file` through `canonical_repo_relative_path`.

Adds `tests/test_repo_paths.py` (26 cases): the base contract it shares with `repo_workspace.canonical_repo_path`, non-`str` input, and the laundering guard that makes `/etc/passwd` return `None` rather than the perfectly-valid-looking `etc/passwd`.

**On the "the two functions agree" claim — corrected, not restored.** The divergence is *intentional and already documented on the function itself*:

> PLUS one stricter rule the write jail does not need but the read jail does: a segment with a trailing space or dot is rejected outright…

So the mirror is of the **base** rule, not of every rule. The docstring now says that, and the new tests pin both halves separately rather than asserting a false equivalence.

**2. Six source comments across three packages justify a decision by citing `harness/`.**

`agentic/config.py` (×2), `agentic/registry.py`, `guardrails/config.py`, `guardrails/call_inventory.py`, `guardrails/tool_broker.py`, `agentic/deepagent_github/repo_workspace.py`. Each points at a file that no longer exists as the *reason* for a duplication or isolation constraint — `agentic/config.py`'s `_LOOPBACK_HOSTS` duplication in particular was justified by a rule with no remaining counterparty.

Each correction keeps the decision's real reasoning and drops only the dead citation. **The duplicated constants stay** — they are still independently justified by out-of-band isolation. `guardrails/config.py` and `agentic/config.py` now cross-reference *each other*, which is true today and verifiable by a reader.

`utils/ops_runner.py`'s two "so the harness route can refuse a shape early" comments now state plainly that **no caller does this today**, leaving `_real_repo_run_timeout_sec`'s `min()` as what actually applies the cap. The symbols are left **public and unrenamed** on purpose: both are still consumed at line 155, so renaming would be churn dressed up as dead-code removal, and would drag the existing test with it.

**Invariant / Governance Impact:** None. Every change outside the new test file is comment/docstring text — no executable statement is added, removed, or reordered. I6 isolation is unchanged and re-verified (`tests/test_agentic_isolation.py` passes; `invariant-guard` 46 passed / 0 failed). The new test imports only `utils.repo_paths`, adding no import edge.

---

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Invariant / Governance refinement

**Scope note:** Out-of-band + utils comments, plus one new test file. No runtime behavior change.

---

## Benefits / why

- Puts a security-relevant validator under test. `canonical_repo_relative_path` decides whether an operator-declared `--read-file` path escapes the clone root; it had no regression net at all, so a future edit could have relaxed the `..`/absolute/`-`-prefix rules silently.
- Restores auditability of three duplication decisions. "We duplicate this because I6 forbids importing X" is only checkable if X exists; pointing at live counterparts makes each one verifiable again.
- Stops `ops_runner` promising an early refusal that nothing performs — a reader currently sees a documented safeguard and has no way to learn it is unreachable.
- No coverage-config change needed: `utils` is already a `--cov` source in `pyproject.toml`, and new test files auto-discover, so `ci.yml` is untouched (and so is any other PR's file).

---

## Risks to monitor

- `tests/test_repo_paths.py` asserts the **current** stricter-than-the-write-jail behavior. That strictness is deliberate today; if someone later decides the read jail should match the write jail exactly, this test is the thing that will fail — and it should be updated deliberately, not deleted, same posture as the Rule 5 tripwire.
- The docstring now points at `tests/test_repo_paths.py` by name. If that file is ever renamed, the citation goes stale in exactly the way this PR is fixing — worth keeping in mind as the general hazard of naming a test in a docstring.
- Comment-only edits are invisible to the test suite by construction, so nothing here is protected by CI. The compensating control is that no executable line moved; `git diff` shows only comment bodies outside the new test.

---

## Checklist

- [x] This change preserves all 6 security invariants and I6 module isolation — comment-only outside the new test; `invariant-guard` 46 passed / 0 failed; `test_agentic_isolation` green
- [x] Touched-module suites pass: `test_repo_paths` (new, 26), `test_ops_runner`, `test_agentic_config`, `test_agentic_repo_workspace`, `test_agentic_isolation`, `test_tool_broker_adversarial` → **340 passed**; guardrails/registry selection → **303 passed, 6 skipped**
- [x] `ruff check --select F,B,S` (CI-blocking) clean on all 9 changed files; broader `E,F,I,B,C4,UP,S` also clean
- [x] New test is deterministic — no network, no clock, no live service, no fixture beyond `pytest.mark.parametrize`
- [x] No `ci.yml` / coverage-config edit required (`utils` already a `--cov` source; test files auto-discover)
- [x] No new external network dependencies introduced
- [x] Commit message follows the title prefix convention

---

## Further comments

**ELI5:** Part of CyClaw was deleted a few days ago. A little file that checks "is this file path trying to escape the folder it's allowed to read?" was left behind with a note saying "there's a test proving I stay in sync" — except that test was deleted too, so the file had *no* tests at all despite still doing real security work. This adds those tests. Separately, several comments around the codebase explain "we copy this value instead of importing it because of that deleted thing" — those now point at things that actually exist, so the next person can check whether the reasoning still holds instead of chasing a ghost.

**No invariant matrix needed** — no executable line changed outside a new, isolated test file.

## Merge order

- **Independent** — no stack parent, base is `main`.
- This run opened 3 PRs from `origin/main` with **no shared files** between them, so they merge in any order. Prefer lowest PR number → highest.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MceD6ySwqCckKPNvwm8JKs

---
_Generated by [Claude Code](https://claude.ai/code/session_01MceD6ySwqCckKPNvwm8JKs)_

### CODE REVIEW

**Comment by cgfixit** [comment]
@codex review

File: tests/test_repo_paths.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        ("docs/guide.md", "docs/guide.md"),
+        ("./README.md", "README.md"),
+        ("docs/./guide.md", "docs/guide.md"),
+        ("docs//guide.md", "docs/guide.md"),
+        # Backslashes are normalized so a Windows-spelled path canonicalizes
+        # to the same string the POSIX spelling produces.
+        ("docs\\guide.md", "docs/guide.md"),
+    ],
+)
+def test_accepts_and_canonicalizes_repo_relative_paths(raw, expected):
```
[the remark this answers is not in the document]
Comment (cgfixit) [reply tests/test_repo_paths.py]: Done in `0133624`.

I checked the claim against the suite rather than taking "repository convention" on faith, and it holds: **1043 of 2559** top-level test functions under `tests/` already carry `-> None`, including annotated parametrize params (`name: str`, `tmp_path: Path`, `source: str`). Adoption is partial, but it's the documented standard in `CLAUDE.md` §5 and `.claude/rules/PROJECT_RULES.md` ("Typing: Fully annotated"), so this file was widening the untyped surface rather than matching the suite.

All five functions and their parametrize params are annotated now.

I took your `object` suggestion for `test_rejects_non_str_input`, with one addition worth calling out — `canonical_repo_relative_path` is typed `(target: str) -> str | None`, so passing `object` is an `arg-type` error at the call site by construction. That test exists *precisely* to exercise the runtime `isinstance` guard against callers who violate that annotation, so the violation is the subject under test:

```python
def test_rejects_non_str_input(raw: object) -> None:
    # `object`, not `str`: the point of this test is the runtime isinstance
    # guard, which exists for callers that violate the annotation. The ignore
    # marks that violation as the subject under test rather than an oversight.
    assert canonical_repo_relative_path(raw) is None  # type: ignore[arg-type]
```

A bare ignore with no explanation would read as someone silencing an inconvenient error, which is the opposite of what it means here.

Verified: 26 tests pass, `ruff check --select F,B,S` and repo-default `ruff check` clean, and `mypy --strict --python-version 3.12 --explicit-package-bases tests/test_repo_paths.py` → *Success: no issues found in 1 source file*.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: agentic/config.py [hunk 2]
```diff
@@ -135,9 +135,10 @@ def _validate_bool(value: object, field_name: str) -> None:
         )
 
 
-# Mirrors harness/ollama.py's own list. Duplicated rather than imported because
-# I6 forbids agentic/ and harness/ importing each other at all; the values are
-# a closed set (the three spellings of localhost), not a tunable.
+# Duplicated rather than imported: I6 forbids agentic/ importing the core
+# request path, and guardrails/config.py keeps its own copy for the same
+# reason; the values are a closed set (the three spellings of localhost),
```
[the remark this answers is not in the document]
Comment (cgfixit) [reply agentic/config.py]: Correct, and a fair hit — fixed in `852cc9b`. This PR's whole premise is that comments should cite things that are true, and the previous pass swapped a stale citation (`harness/ollama.py`, deleted in #1367) for an **inaccurate** one.

Verified both of your claims against the tree:

1. **I6 doesn't cover it.** `invariant-guard`'s `CORE_FILES` is exactly `("gate.py", "gate_ops.py", "gate_auth.py", "gate_memory.py", "graph.py", "mcp_hybrid_server.py")`. `llm/client.py` — which holds the identical `frozenset` at line 92 — is not among them, so I6 does not forbid `agentic/` from importing it. My comment named the wrong invariant.
2. **`agentic` and `guardrails` are not mutually isolated.** `agentic/context.py:20`, `agentic/registry.py:37`, `agentic/fsconnect/client.py:23,26`, and `agentic/harness_optimizer/governance.py:12` all import `guardrails.rails`.

The real reasons, both checkable rather than asserted:

- **Out-of-band convention** — `grep -rn "^from llm\|^import llm" agentic/ guardrails/ sync/ telegram/ opentweet/` returns **nothing**. No out-of-band package imports `llm/` today.
- **Import cost** — `llm/client.py` pulls `httpx`, `yaml`, `utils.config_validation`, `utils.errors` and `utils.spend` at module level. Importing it for three strings drags all of that into `agentic`'s import graph.

The comment now states both, and says explicitly that it is *not* an I6 constraint, so the next reader doesn't re-derive the same wrong claim a third time.

One partial disagreement, for the record: `guardrails/config.py` is left as-is deliberately. Its wording is pre-existing and reads "that module is the core request path, and guardrails must not import it (**out-of-band isolation**)" — it never invoked I6, so it wasn't repeating the bad rationale. The only thing my earlier commit changed there was the dangling `harness/ollama.py` cross-reference. Rewriting otherwise-accurate pre-existing prose would widen a scoped cleanup.

I also didn't take the "shared allowed helper" option: a new module to host three localhost spellings is more coupling than the duplication costs, and this PR is explicitly a comment-accuracy pass, not a refactor.

Verified: `test_agentic_config.py` + `test_agentic_isolation.py` → 82 passed, `ruff` clean both ways, `invariant-guard` 46 passed / 0 failed. Comment-only, no executable line moved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: utils/repo_paths.py [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
+reject the same escapes *before* a staged path is forwarded as ``--read-file``.
 
-This module is a stdlib-only mirror of that acceptance rule. A drift test in
-``tests/test_harness_agent_routes.py`` asserts the two functions agree on a
-matrix of safe and hostile inputs so a future jail change cannot leave the
-control plane accepting what the clone will skip (or vice versa).
+This module is a stdlib-only mirror of that acceptance rule, plus one
+deliberately stricter segment check documented on the function itself.
+``tests/test_repo_paths.py`` pins both halves: the shared base contract, and
+the trailing dot/space rule the write jail does not need.
```
[the remark this answers is not in the document]
Comment (cgfixit) [reply utils/repo_paths.py]: Right again — fixed in `240eda4`. This is the sharpest of the three findings on this PR: I restored *unit* coverage and then described it as pinning "the shared base contract," which it wasn't doing. A mirror checked only against its own hard-coded expectations cannot notice the original moving, which is exactly the regression the deleted test existed to catch.

**Verified the divergence empirically before asserting anything.** The two functions are byte-identical except one clause — `repo_paths` adds `part != part.rstrip(" .")`. Running both over the matrix:

```
--- base contract (17 inputs) ---   disagreements: 0
--- trailing dot/space ---
  'README.md.'     read=None  write='README.md.'
  'README.md '     read=None  write='README.md '
  'docs./guide.md' read=None  write='docs./guide.md'
  'docs /guide.md' read=None  write='docs /guide.md'
```

Two new tests, taking your split:

- `test_both_jails_agree_on_the_shared_base_contract` — 17 inputs (accepted spellings plus absolute, drive-qualified, traversal, leading-dash, colon-in-segment, NUL, empty, dot-only) asserting `canonical_repo_relative_path(raw) == canonical_repo_path(raw)`.
- `test_read_jail_is_stricter_..._only_on_trailing_dot_or_space` — pins the deliberate divergence in **both** directions: read jail `None`, write jail still accepting. That second assertion is what keeps the gap exactly one rule wide instead of silently widening.

**I also checked the pin isn't vacuous**, since a comparison test that can't fail is worse than none. Monkeypatching `canonical_repo_path` in memory:

```
after mutating write jail: read='README.md' write='MUTATED' -> test would FAIL (good)
```

On your isolation note — agreed, and worth stating plainly: the test file importing `agentic` is fine because I6 constrains the six core modules, not tests (`tests/test_agentic_*.py` already do this). The runtime isolation is exactly *why* the rule is duplicated, and therefore exactly why a test has to hold the two in step. I updated `utils/repo_paths.py`'s module docstring to say that rather than just claiming coverage exists.

Verified: **47 tests pass** (was 26), `ruff` clean both ways, `test_agentic_isolation` still green. `mypy --strict` reports **0 errors in this file**; the transitive `agentic/` legacy noise is pre-existing — an untouched `tests/test_agentic_repo_workspace.py` reports 348 of the same, per `CLAUDE.md` §4's note that the tree doesn't pass clean end-to-end.

---
_Generated by [Claude Code](https://claude.ai/code)_
